runtime/lua/vim/_range.lua

```diff
@@ -0,0 +1,85 @@
+local pos = require('vim._pos')
+
+---@class vim.Range
+---@field start vim.Pos Start position.
+---@field end_ vim.Pos End position, exclusive.
+---@overload fun(start: vim.Pos, end_: vim.Pos): vim.Range
+---@overload fun(start_row: integer, start_col: integer, end_row: integer, end_col: integer): vim.Range
+local Range = {}
+Range.__index = Range
+
+---@package
+---@overload fun(start: vim.Pos, end_: vim.Pos): vim.Range
+---@overload fun(start_row: integer, start_col: integer, end_row: integer, end_col: integer): vim.Range
+function Range:new(...)
+  local args = { ... }
+  ---@type vim.Pos, vim.Pos
+  local start, end_
+
+  if #args == 2 then
+    ---@cast args [vim.Pos, vim.Pos]
+    start, end_ = unpack(args)
+  elseif #args == 4 then
+    ---@cast args [integer, integer, integer, integer]
+    start, end_ = pos(unpack(args, 1, 2)), pos(unpack(args, 3, 4))
+  else
+    assert(false)
+  end
+
+  ---@class vim.Range
+  self = setmetatable({}, self)
+  self.start = start
+  self.end_ = end_
+  return self
+end
+
+---@param a vim.Range
+---@param b vim.Range
+function Range.contains(a, b)
+  return a.start <= b.start and a.end_ >= b.end_
+end
+
+---@param a vim.Range
+---@param b vim.Range
+function Range.intercepts(a, b)
+  return a.end_ > b.start and a.start < b.end_
```

Comment:
docstring will help. unlike "contains" (`has()`), it looks like this one checks if they "touch" but do not fully overlap.